### PR TITLE
BUG: f2py:  make callbacks threadsafe

### LINKED ISSUE #16513: BUG: f2py produces wrapper code that are not threadsafe
f2py produces wrapper code that are not threadsafe and can segfault. See https://github.com/scipy/scipy/issues/9658

### Reproducing code example:

The segfault in the code below is caused by the f2py wrapper.

```python
import concurrent.futures
import time
import scipy.optimize

def objective1(x):
    print(x, "1")
    time.sleep(0.1)
    return x[0]**2

def objective2(x):
    print(x, "2")
    time.sleep(0.1)
    return (x[0]-1)**2

min_method = "COBYLA"

def minimizer1():
    return scipy.optimize.minimize(objective1,
                                   [0.0],
                                   method=min_method)

def minimizer2():
    return scipy.optimize.minimize(objective2,
                                   [0.0],
                                   method=min_method)

with concurrent.futures.ThreadPoolExecutor() as pool:
    tasks = []
    tasks.append(pool.submit(minimizer1))
    tasks.append(pool.submit(minimizer2))

    for t in tasks:
        print(t.result())
```

### Error message:

#0  0x00007f5e94d24dc0 in f2py_rout.cobyla_minimize () from /home/XXXX/lib/python3.7/site-packages/scipy/optimize/_cobyla.cpython-37m-x86_64-linux-gnu.so
#1  0x000055b6ce95146b in _PyObject_FastCallKeywords (callable=0x7f5e94d73670, stack=0x7f5e88001680, nargs=1, kwnames=0x7f5e94d74a08) at /tmp/build/80754af9/python_1546061345851/work/Objects/call.c:199
#2  0x000055b6ce9ac378 in call_function (kwnames=0x7f5e94d74a08, oparg=<optimized out>, pp_stack=<synthetic pointer>) at /tmp/build/80754af9/python_1546061345851/work/Python/ceval.c:4619
[....]

### PR DESCRIPTION
**Pull request by pv**
Make f2py Python callback functions threadsafe by putting the associated globals to thread-local storage. Use compiler TLS if possible (implemented for gcc & msvc), and fall back to Python-provided TLS otherwise.

Fixes gh-16513

### CODE REVIEW

File: numpy/f2py/cb_rules.py [hunk 1]
```diff
@@ -33,114 +33,151 @@
...
-int #name#_nofargs = 0;
-jmp_buf #name#_jmpbuf;
+typedef struct {
+    PyObject *capi;
+    PyTupleObject *args_capi;
+    int nofargs;
+    jmp_buf jmpbuf;
+} #name#_t;
+
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4))) && !defined(F2PY_USE_PYTHON_TLS)
```
Comment (charris) [inline numpy/f2py/cb_rules.py:43]: Would be nice to break this line.

File: numpy/f2py/cb_rules.py [hunk 2]
```diff
@@ -33,114 +33,151 @@
...
+static __thread #name#_t *_active_#name# = NULL;
+static #name#_t *swap_active_#name#(#name#_t *ptr) {
+    #name#_t *prev = _active_#name#;
+    _active_#name# = ptr;
+    return prev;
+}
```
Comment (charris) [inline numpy/f2py/cb_rules.py:49]: Blank lines around function definition.

File: numpy/f2py/cb_rules.py [hunk 3]
```diff
@@ -33,114 +33,151 @@
...
+    return _active_#name#;
+}
+#elif defined(_MSC_VER) && !defined(F2PY_USE_PYTHON_TLS)
+static __declspec(thread) #name#_t *_active_#name# = NULL;
+static #name#_t *swap_active_#name#(#name#_t *ptr) {
+    #name#_t *prev = _active_#name#;
+    _active_#name# = ptr;
+    return prev;
+}
+static #name#_t *get_active_#name#() {
```
Comment (charris) [inline numpy/f2py/cb_rules.py:60]: see above.

Comment (charris) [reply numpy/f2py/cb_rules.py:60]: Does that missing arguments need `void`?

File: numpy/f2py/cb_rules.py [hunk 4]
```diff
@@ -33,114 +33,151 @@
...
+    return _active_#name#;
+}
+#else
+static #name#_t *swap_active_#name#(#name#_t *ptr) {
+    char *key = "__f2py_cb_#name#";
+    return (#name#_t *)F2PySwapThreadLocalCallbackPtr(key, ptr);
+}
+static #name#_t *get_active_#name#() {
```
Comment (charris) [inline numpy/f2py/cb_rules.py:68]: See above.

File: numpy/f2py/cb_rules.py [hunk 5]
```diff
@@ -33,114 +33,151 @@
...
 capi_fail:
-\tfprintf(stderr,\"Call-back #name# failed.\\n\");
-\tPy_XDECREF(capi_return);
-\tPy_XDECREF(capi_arglist_list);
-\tif (capi_longjmp_ok)
-\t\tlongjmp(#name#_jmpbuf,-1);
+    fprintf(stderr,\"Call-back #name# failed.\\n\");
+    Py_XDECREF(capi_return);
+    Py_XDECREF(capi_arglist_list);
+    if (capi_longjmp_ok)
```
Comment (charris) [inline numpy/f2py/cb_rules.py:177]: Use `{}` here.

**Comment by charris** [comment]
The formatting cleanup is nice. I wouldn't complain if you made it a separate PR and put the rest of this on top after it goes in. There is still more cleanup to do, but I think others can take care of it.

**Comment by pv** [comment]
Style changes in gh-16520 separately.

File: numpy/f2py/cb_rules.py (hunk 3 above)
Comment (pv) [reply numpy/f2py/cb_rules.py]: Yes, fixed (and fixed also the othe things).

**Comment by pv** [comment]
Rebased.

**Comment by charris** [comment]
@pv IIRC, swapping was the main reason we could not use the newer interfaces that are in the include files. If you have any ideas for an public swap procedure I'd like to hear them.

@melissawm There are lots of opportunity for additional code beautification here, much of it whitespace :)

**Comment by sturlamolden** [comment]
With C11 or C++11 we can use ```thread_local``` macro from the standard library, defined in ```threads.h```. That will work on any standards compliant C or C++ compiler, even if The non-standard ```__thread``` and ```__declspec(thread)``` extensions are not defined.

**Comment by charris** [comment]
Let's put this in, thanks Pauli. I'm a bit undecided on backporting, I'd rather keep changes to a minimum for the 1.19 release.

**Comment by pv** [comment]
@charris: hmm, TBH I'm not sure what you mean by swapping and what are the newer interfaces here.

**Comment by charris** [comment]
I'm talking about the inline functions that replaced macros in order to hide ndarray internals. I'm not sure that is what this is about, but I recall data pointers being swapped somewhere in f2py and that kept us from an easy path to using
```
#define NPY_NO_DEPRECATED_API NPY_API_VERSION
```

Now that I think about it, it would be good to have a release note for this, maybe under improvements.

**Comment by pv** [comment]
Ok, added release note and Sturla's suggestion in #16524

**Comment by seberg** [comment]
Sorry, just curious... Is this approach fully safe with restoring always the same state? I noticed there is a `longjmp` in there. And we still have some annoying issues in the fft code with `longjmp` not being restored.

It seems to me however, that we may have to make it a doubly-linked list to really always correctly restore things if multiple threads finish in random order?  But maybe I just need to think about it longer...

**Comment by pv** [comment]
It will unwind in stack order, and each thread has its own (stack and threadlocal global).

**Comment by pv** [comment]
Ie it should not matter what other threads are doing.

**Comment by seberg** [comment]
Ah, nvm. totally different situation since its on the stack here.

**Pull request merged**